lib/hotel.rb

```diff
@@ -0,0 +1,189 @@
+
+    def make_reservation(check_in, check_out, room_num)
+      if room_availability(check_in, check_out).include?(room_num)
+        @reservation_made = Reservation.new(check_in, check_out, room_num)
+        @reservation_collection << @reservation_made
+      else
+        raise ArgumentError.new("It appears the room you requested is booked during that date range")
+      end
+      # @reservation_collection << @reservation_made
+    end
+
+
+    def make_block_reservation(check_in, check_out, num_of_rooms)
+      available_rooms = room_availability(check_in, check_out)
+      if available_rooms.length >= num_of_rooms
+        block_reservation = BlockRoom.new(check_in, check_out, num_of_rooms)
+        block_reservation.block_of_rooms << available_rooms.pop(num_of_rooms)
+        @block_reservation_collection << block_reservation
+        return block_reservation.block_of_rooms
+      else
+        return raise ArgumentError.new("We do not have enough rooms to reserve a block")
+      end
+    end
+
+    # ability to reserve a room within a block of rooms
+    #needs to match the date range of the block
+    def reserve_room_in_block(check_in, check_out, num_of_rooms)
+
+      check_in = Date.parse(check_in)
+      check_out = Date.parse(check_out)
+
+      @block_reservation_collection.each do |entry|
+        if check_in == entry.check_in && check_out == entry.check_out
+          if entry.available_rooms[0].length >= num_of_rooms
+            num_of_rooms.times do
+              entry.booked_rooms << entry.available_rooms[0].pop
+            end
+            return entry.booked_rooms
+          end
+        end
+      end
+    end
+
+    def blocked_rooms_availability(check_in, check_out)
+      check_in = Date.parse(check_in)
+      check_out = Date.parse(check_out)
+
+      @block_reservation_collection.each do |entry|
+        if check_in == entry.check_in && check_out == entry.check_out
+          if entry.available_rooms[0].length >= 0
+            return true
+          else false
+          end
+        end
+      end
+    end
+
+    def date_list_of_reservations(date)
+      date_list = []
+      date = Date.parse(date)
+      @reservation_collection.each do |entry|
+        # binding.pry
+        if entry.check_in <= date && entry.check_out >= date
+          date_list << [entry.room_num, entry.check_in.to_s, entry.check_out.to_s]
+        end
+      end
+      return date_list
+    end #def
+
+    def room_availability(check_in, check_out)
+      #can be string with dates
+      rooms_available = @rooms.clone
+
+      # [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]
+      check_in = Date.parse(check_in)
+      check_out = Date.parse(check_out)
+
+      if check_in > check_out
+        raise ArgumentError.new("Invalid date range")
+      end
+
+      @reservation_collection.each do |entry|
+        if check_in < entry.check_in && check_in < entry.check_out && check_out < entry.check_out && check_out > entry.check_in
```

Comment:
Woah, this logic is super complicated! I have a few thoughts about it:

- You ought to be able to take advantage of the bounds checking in `DateRange` to simplify this logic, since you know `check_in < check_out`
- This is pretty hard to read - my eyes immediately start to glaze over just looking at it. Might be a good place to leave a big comment explaining what you intended.
- You have almost exactly the same logic repeated a little lower for room blocks.
- This logic seems like it's more related to a `DateRange` than it is to a `Hotel`. Maybe you could add it there as an instance method `DateRange#overlap?(checkin, checkout)`. Then this big complex block could be shortened to
    ```ruby
    if entry.date_range.overlap?(check_in, check_out)
      rooms_available.delete(entry.room_num)
    end
    ```
    That would also let you DRY out the repeated code below.

Reply:
Making it an instance method would also make it way easier to test!